### PR TITLE
CP-42064: Fix storage migration for NBD-backed storage

### PR DESCRIPTION
**Pull request by robhoes**

### CODE REVIEW

File: ocaml/xapi/storage_migrate.ml [hunk 1]
```diff
@@ -323,12 +323,12 @@ module Local = StorageAPI (Idl.Exn.GenClient (struct
...
-  let xendisks, _, _, _ =
+  let _, blockdevices, _, nbds =
     Storage_interface.implementations_of_backend backend
   in
-  match xendisks with
-  | xendisk :: _ -> (
-      let path = xendisk.Storage_interface.params in
+  match (blockdevices, nbds) with
+  | blockdevice :: _, _ -> (
+      let path = blockdevice.Storage_interface.path in
```
Comment (psafont) [inline ocaml/xapi/storage_migrate.ml:331]: This change is surprising, did `implementation_of_backend` return duplicated information previously?

File: ocaml/xapi/storage_migrate.ml [hunk 2]
```diff
@@ -360,23 +371,32 @@ let with_activated_disk ~dbg ~sr ~vdi ~dp f =
...
+      let path_and_nbd =
         Option.map
           (fun (vdi, backend) ->
-            let _, blockdevs, files, _ =
+            let _xendisks, blockdevs, files, nbds =
               Storage_interface.implementations_of_backend backend
             in
-            match (blockdevs, files) with
-            | {path} :: _, _ | _, {path} :: _ ->
+            match (files, blockdevs, nbds) with
```
Comment (psafont) [inline ocaml/xapi/storage_migrate.ml:380]: blockdevs and files got swapped, but the match remains the same, this doesn't look intentional

File: ocaml/xapi/storage_migrate.ml (hunk 1 above)
Comment (robhoes) [reply ocaml/xapi/storage_migrate.ml:331]: Yes, see here: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/storage_smapiv1.ml#L548-L554. The `params`, from the return value of `VDI.attach`, ends up in both places.

File: ocaml/xapi/attach_helpers.ml [hunk 3]
```diff
@@ -132,3 +132,44 @@ let with_vbds rpc session_id __context vm vdis mode f =
             !vbds
       )
     )
+
+module NbdClient = struct
```
Comment (lindig) [inline ocaml/xapi/attach_helpers.ml:136]: Minor thing, but would use `NBDClient` because `NBD` is an acronym.

**Review by psafont — APPROVED** [review]

File: ocaml/xapi/attach_helpers.ml [hunk 4]
```diff
@@ -132,3 +132,44 @@ let with_vbds rpc session_id __context vm vdis mode f =
...
+  let print_fork_error f =
```
Comment (lindig) [inline ocaml/xapi/attach_helpers.ml:137]: Call this `try'n'catch`?

File: ocaml/xapi/attach_helpers.ml [hunk 5]
```diff
@@ -132,3 +132,44 @@ let with_vbds rpc session_id __context vm vdis mode f =
...
+          error "Forkhelpers.Spawn_internal_error(%s, %s, WSIGNALED %d)" stderr
+            stdout n ;
+          raise e
+      | Unix.WSTOPPED n ->
+          error "Forkhelpers.Spawn_internal_error(%s, %s, WSTOPPED %d)" stderr
+            stdout n ;
+          raise e
+    )
+
+  let run_command cmd args =
```
Comment (lindig) [inline ocaml/xapi/attach_helpers.ml:155]: Just `run` or `execute`?

**Review by lindig — APPROVED** [review]

**Pull request merged**